The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16530377205/job/46754105023) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46754105023) [(plain)](https://github.com/rust-lang/rust/commit/3c72881993bacbfbbc07442b42a46eca60b23404/checks/46754105023/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

+ error: free constant item without body
+   --> $DIR/suggest-const-item.rs:127:1
+    |
+ LL | const O<'a>: &'a u64;
+    | ^^^^^^^^^^^^^^^^^^^^-
+    |                     |
+    |                     help: provide a definition for the constant: `= <expr>;`
+ 
1 error: could not determine the target branch for this `#[const_continue]`
-   --> $DIR/suggest-const-item.rs:18:32
+   --> $DIR/suggest-const-item.rs:19:32
3    |
4 LL |                     break 'blk const_fn();
5    |                                ^^^^^^^^^^ this value must be a literal or a monomorphic const

7    = help: try extracting the expression into a `const` item
8 
9 error: could not determine the target branch for this `#[const_continue]`
-   --> $DIR/suggest-const-item.rs:23:32
+   --> $DIR/suggest-const-item.rs:24:32
11    |
12 LL |                     break 'blk const { const_fn() };
13    |                                ^^^^^^^^^^^^^^^^^^^^ `const` blocks may use generics, and are not evaluated early enough

15    = help: try extracting the expression into a `const` item
16 
17 error: could not determine the target branch for this `#[const_continue]`
-   --> $DIR/suggest-const-item.rs:28:32
+   --> $DIR/suggest-const-item.rs:29:32
19    |
20 LL |                     break 'blk N;
21    |                                ^ constant parameters may use generics, and are not evaluated early enough

23    = help: try extracting the expression into a `const` item
24 
25 error: could not determine the target branch for this `#[const_continue]`
-   --> $DIR/suggest-const-item.rs:33:32
+   --> $DIR/suggest-const-item.rs:34:32
27    |
28 LL |                     break 'blk 1 + 1;
29    |                                ^^^^^ this value must be a literal or a monomorphic const

31    = help: try extracting the expression into a `const` item
32 
33 error: could not determine the target branch for this `#[const_continue]`
-   --> $DIR/suggest-const-item.rs:75:36
+   --> $DIR/suggest-const-item.rs:76:36
35    |
36 LL |                         break 'blk Self::N;
37    |                                    ^^^^^^^ this value is too generic

38 
- error: aborting due to 5 previous errors
+ error: this `#[loop_match]` state value has type `U`, which is not supported
+   --> $DIR/suggest-const-item.rs:114:13
+    |
+ LL |             state = 'blk: {
+    |             ^^^^^
+    |
+    = note: only integers, floats, bool, char, and enums without fields are supported
+ 
+ error: this `#[loop_match]` state value has type `&u64`, which is not supported
+   --> $DIR/suggest-const-item.rs:132:9
+    |
+ LL |         state = 'blk: {
+    |         ^^^^^
+    |
+    = note: only integers, floats, bool, char, and enums without fields are supported
+ 
+ error: aborting due to 8 previous errors
---
-   --> /checkout/tests/ui/loop-match/suggest-const-item.rs:132:9
+ error: free constant item without body
+   --> $DIR/suggest-const-item.rs:127:1
+    |
+ LL | const O<'a>: &'a u64;
+    | ^^^^^^^^^^^^^^^^^^^^-
+    |                     |
+    |                     help: provide a definition for the constant: `= <expr>;`
+ 
+   --> $DIR/suggest-const-item.rs:19:32
+   --> $DIR/suggest-const-item.rs:24:32
+   --> $DIR/suggest-const-item.rs:29:32
+   --> $DIR/suggest-const-item.rs:34:32
+   --> $DIR/suggest-const-item.rs:76:36
+ error: this `#[loop_match]` state value has type `U`, which is not supported
+   --> $DIR/suggest-const-item.rs:114:13
+    |
+ LL |             state = 'blk: {
+    |             ^^^^^
+    |
+    = note: only integers, floats, bool, char, and enums without fields are supported
+ 
+ error: this `#[loop_match]` state value has type `&u64`, which is not supported
+   --> $DIR/suggest-const-item.rs:132:9
+    |
+ LL |         state = 'blk: {
+    |         ^^^^^
+    |
+    = note: only integers, floats, bool, char, and enums without fields are supported
+ 
+ error: aborting due to 8 previous errors
---
To only update this specific test, also pass `--test-args loop-match/suggest-const-item.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/loop-match/suggest-const-item.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/loop-match/suggest-const-item" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: free constant item without body
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:127:1
   |
LL | const O<'a>: &'a u64;
   | ^^^^^^^^^^^^^^^^^^^^-
   |                     |
   |                     help: provide a definition for the constant: `= <expr>;`

error: could not determine the target branch for this `#[const_continue]`
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:19:32
   |
LL |                     break 'blk const_fn();
   |                                ^^^^^^^^^^ this value must be a literal or a monomorphic const
   |
   = help: try extracting the expression into a `const` item

error: could not determine the target branch for this `#[const_continue]`
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:24:32
   |
LL |                     break 'blk const { const_fn() };
   |                                ^^^^^^^^^^^^^^^^^^^^ `const` blocks may use generics, and are not evaluated early enough
   |
   = help: try extracting the expression into a `const` item

error: could not determine the target branch for this `#[const_continue]`
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:29:32
   |
LL |                     break 'blk N;
   |                                ^ constant parameters may use generics, and are not evaluated early enough
   |
   = help: try extracting the expression into a `const` item

error: could not determine the target branch for this `#[const_continue]`
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:34:32
   |
LL |                     break 'blk 1 + 1;
   |                                ^^^^^ this value must be a literal or a monomorphic const
   |
   = help: try extracting the expression into a `const` item

error: could not determine the target branch for this `#[const_continue]`
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:76:36
   |
LL |                         break 'blk Self::N;
   |                                    ^^^^^^^ this value is too generic

error: this `#[loop_match]` state value has type `U`, which is not supported
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:114:13
   |
LL |             state = 'blk: {
   |             ^^^^^
   |
   = note: only integers, floats, bool, char, and enums without fields are supported

error: this `#[loop_match]` state value has type `&u64`, which is not supported
##[error]  --> /checkout/tests/ui/loop-match/suggest-const-item.rs:132:9
   |
LL |         state = 'blk: {
   |         ^^^^^
   |
   = note: only integers, floats, bool, char, and enums without fields are supported
```

</details>

--- a/compiler/rustc_mir_build/messages.ftl
+++ b/compiler/rustc_mir_build/messages.ftl
@@ -86,9 +86,15 @@ mir_build_confused = missing patterns are not covered because `{$variable}` is i
 
 mir_build_const_continue_bad_const = could not determine the target branch for this `#[const_continue]`
     .label = this value is too generic
-    .note = the value must be a literal or a monomorphic const
 
 mir_build_const_continue_missing_value = a `#[const_continue]` must break to a label with a value
+
+mir_build_const_continue_not_const = could not determine the target branch for this `#[const_continue]`
+    .help = try extracting the expression into a `const` item
+
+mir_build_const_continue_not_const_const_block = `const` blocks may use generics, and are not evaluated early enough
+mir_build_const_continue_not_const_const_other = this value must be a literal or a monomorphic const
+mir_build_const_continue_not_const_constant_parameter = constant parameters may use generics, and are not evaluated early enough
 
 mir_build_const_continue_unknown_jump_target = the target of this `#[const_continue]` is not statically known
     .label = this value must be a literal or a monomorphic const

--- a/compiler/rustc_mir_build/src/builder/scope.rs
+++ b/compiler/rustc_mir_build/src/builder/scope.rs
@@ -100,7 +100,9 @@ use tracing::{debug, instrument};
 
 use super::matches::BuiltMatchTree;
 use crate::builder::{BlockAnd, BlockAndExtension, BlockFrame, Builder, CFG};
-use crate::errors::{ConstContinueBadConst, ConstContinueUnknownJumpTarget};
+use crate::errors::{
+    ConstContinueBadConst, ConstContinueNotMonomorphicConst, ConstContinueUnknownJumpTarget,
+};
 
 #[derive(Debug)]
 pub(crate) struct Scopes<'tcx> {
@@ -867,7 +869,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             span_bug!(span, "break value must be a scope")
         };
 
-        let constant = match &self.thir[value].kind {
+        let expr = &self.thir[value];
+        let constant = match &expr.kind {
             ExprKind::Adt(box AdtExpr { variant_index, fields, base, .. }) => {
                 assert!(matches!(base, AdtExprBase::None));
                 assert!(fields.is_empty());
@@ -887,7 +890,27 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     ),
                 }
             }
-            _ => self.as_constant(&self.thir[value]),
+
+            ExprKind::Literal { .. }
+            | ExprKind::NonHirLiteral { .. }
+            | ExprKind::ZstLiteral { .. }
+            | ExprKind::NamedConst { .. } => self.as_constant(&self.thir[value]),
+
+            other => {
+                use crate::errors::ConstContinueNotMonomorphicConstReason as Reason;
+
+                let span = expr.span;
+                let reason = match other {
+                    ExprKind::ConstParam { .. } => Reason::ConstantParameter { span },
+                    ExprKind::ConstBlock { .. } => Reason::ConstBlock { span },
+                    _ => Reason::Other { span },
+                };
+
+                self.tcx
+                    .dcx()
+                    .emit_err(ConstContinueNotMonomorphicConst { span: expr.span, reason });
+                return block.unit();
+            }
         };
 
         let break_index = self

--- a/compiler/rustc_mir_build/src/errors.rs
+++ b/compiler/rustc_mir_build/src/errors.rs
@@ -1214,6 +1214,38 @@ pub(crate) struct LoopMatchArmWithGuard {
 }
 
 #[derive(Diagnostic)]
+#[diag(mir_build_const_continue_not_const)]
+#[help]
+pub(crate) struct ConstContinueNotMonomorphicConst {
+    #[primary_span]
+    pub span: Span,
+
+    #[subdiagnostic]
+    pub reason: ConstContinueNotMonomorphicConstReason,
+}
+
+#[derive(Subdiagnostic)]
+pub(crate) enum ConstContinueNotMonomorphicConstReason {
+    #[label(mir_build_const_continue_not_const_constant_parameter)]
+    ConstantParameter {
+        #[primary_span]
+        span: Span,
+    },
+
+    #[label(mir_build_const_continue_not_const_const_block)]
+    ConstBlock {
+        #[primary_span]
+        span: Span,
+    },
+
+    #[label(mir_build_const_continue_not_const_const_other)]
+    Other {
+        #[primary_span]
+        span: Span,
+    },
+}
+
+#[derive(Diagnostic)]
 #[diag(mir_build_const_continue_bad_const)]
 pub(crate) struct ConstContinueBadConst {
     #[primary_span]

--- a/tests/ui/loop-match/suggest-const-item.rs
+++ b/tests/ui/loop-match/suggest-const-item.rs
@@ -1,0 +1,174 @@
+#![allow(incomplete_features)]
+#![feature(loop_match)]
+#![feature(generic_const_items)]
+#![crate_type = "lib"]
+
+const fn const_fn() -> i32 {
+    1
+}
+
+#[unsafe(no_mangle)]
+fn suggest_const_block<const N: i32>() -> i32 {
+    let mut state = 0;
+    #[loop_match]
+    loop {
+        state = 'blk: {
+            match state {
+                0 => {
+                    #[const_continue]
+                    break 'blk const_fn();
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+                1 => {
+                    #[const_continue]
+                    break 'blk const { const_fn() };
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+                2 => {
+                    #[const_continue]
+                    break 'blk N;
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+                _ => {
+                    #[const_continue]
+                    break 'blk 1 + 1;
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+            }
+        }
+    }
+    state
+}
+
+struct S;
+
+impl S {
+    const M: usize = 42;
+
+    fn g() {
+        let mut state = 0;
+        #[loop_match]
+        loop {
+            state = 'blk: {
+                match state {
+                    0 => {
+                        #[const_continue]
+                        break 'blk Self::M;
+                    }
+                    _ => panic!(),
+                }
+            }
+        }
+    }
+}
+
+trait T {
+    const N: usize;
+
+    fn f() {
+        let mut state = 0;
+        #[loop_match]
+        loop {
+            state = 'blk: {
+                match state {
+                    0 => {
+                        #[const_continue]
+                        break 'blk Self::N;
+                        //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                    }
+                    _ => panic!(),
+                }
+            }
+        }
+    }
+}
+
+impl T for S {
+    const N: usize = 1;
+}
+
+impl S {
+    fn h() {
+        let mut state = 0;
+        #[loop_match]
+        loop {
+            state = 'blk: {
+                match state {
+                    0 => {
+                        #[const_continue]
+                        break 'blk Self::N;
+                    }
+                    _ => panic!(),
+                }
+            }
+        }
+    }
+}
+
+trait T2<U> {
+    const L: u32;
+
+    fn p() {
+        let mut state = 0;
+        #[loop_match]
+        loop {
+            state = 'blk: {
+                match state {
+                    0 => {
+                        #[const_continue]
+                        break 'blk Self::L;
+                        //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                    }
+                    _ => panic!(),
+                }
+            }
+        }
+    }
+}
+
+const SIZE_OF<T>: usize = size_of::<T>();
+
+fn q<T>() {
+    let mut state = 0;
+    #[loop_match]
+    loop {
+        state = 'blk: {
+            match state {
+                0 => {
+                    #[const_continue]
+                    break 'blk SIZE_OF::<T>;
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+                _ => panic!(),
+            }
+        }
+    }
+}
+
+trait Trait<T> {
+    const X: usize = 9000;
+    const Y: usize = size_of::<T>();
+}
+
+impl<T> Trait<T> for () {}
+
+fn r<T>() {
+    let mut state = 0;
+    #[loop_match]
+    loop {
+        state = 'blk: {
+            match state {
+                0 => {
+                    #[const_continue]
+                    break 'blk <() as Trait<T>>::X;
+                }
+                1 => {
+                    #[const_continue]
+                    break 'blk <() as Trait<T>>::Y;
+                    //~^ ERROR could not determine the target branch for this `#[const_continue]`
+                }
+                _ => panic!(),
+            }
+        }
+    }
+}

--- a/tests/ui/loop-match/suggest-const-item.stderr
+++ b/tests/ui/loop-match/suggest-const-item.stderr
@@ -1,0 +1,58 @@
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:19:32
+   |
+LL |                     break 'blk const_fn();
+   |                                ^^^^^^^^^^ this value must be a literal or a monomorphic const
+   |
+   = help: try extracting the expression into a `const` item
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:24:32
+   |
+LL |                     break 'blk const { const_fn() };
+   |                                ^^^^^^^^^^^^^^^^^^^^ `const` blocks may use generics, and are not evaluated early enough
+   |
+   = help: try extracting the expression into a `const` item
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:29:32
+   |
+LL |                     break 'blk N;
+   |                                ^ constant parameters may use generics, and are not evaluated early enough
+   |
+   = help: try extracting the expression into a `const` item
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:34:32
+   |
+LL |                     break 'blk 1 + 1;
+   |                                ^^^^^ this value must be a literal or a monomorphic const
+   |
+   = help: try extracting the expression into a `const` item
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:76:36
+   |
+LL |                         break 'blk Self::N;
+   |                                    ^^^^^^^ this value is too generic
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:119:36
+   |
+LL |                         break 'blk Self::L;
+   |                                    ^^^^^^^ this value is too generic
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:139:32
+   |
+LL |                     break 'blk SIZE_OF::<T>;
+   |                                ^^^^^^^^^^^^ this value is too generic
+
+error: could not determine the target branch for this `#[const_continue]`
+  --> $DIR/suggest-const-item.rs:167:32
+   |
+LL |                     break 'blk <() as Trait<T>>::Y;
+   |                                ^^^^^^^^^^^^^^^^^^^ this value is too generic
+
+error: aborting due to 8 previous errors
+

